### ISSUE TITLE
*Request unsafe in presence of memory leaks

### ISSUE THREAD

**Issue by Rufflewind** [bsteinb/rsmpi#17]
~~~rust
{
    let buffer = vec![0u8; 512];
    let request = dest.immediate_send(&buffer as &[u8]);
    std::mem::forget(request);
}
// request is still alive but buf is dead
~~~

a.k.a. the [JoinGuard problem](https://github.com/rust-lang/rust/issues/24292).

This could perhaps be fixed by using a [Scope-based mechanism](https://aturon.github.io/crossbeam-doc/crossbeam/struct.Scope.html), or by asking for an owned buffer instead.

**Comment by bsteinb** [comment]
Good catch, thank you.

Although the scope-based mechanism does lead to somewhat convoluted code, I think it is the more general solution. This way the non-blocking operations can still be used on borrowed buffers and thus the same buffer can be used in more than one send operation, or can be involved in a send operation and read from at the same time. Using the existing guard based design with owned buffers would likely require `Arq` and friends to make this work, not sure one could get the C code that actually fulfils the request to uphold the exclusivity constraints though.

I still have to think some more about whether it is possible to carry over the freedom of when exactly one wants to wait for a specific request (or even cancel it) into the scope based design. Only being able to wait for all operations initiated within a scope at its exit seems like a restriction. But then again, this might be what is necessary to actually provide a safe API in the Rust sense of the word.

Currently, I do not have huge amounts of free time on my hands, would you be willing to do some work on this issue?

**Comment by Rufflewind** [comment]
Can't promise I'll have time to work on it either, but I'll think about it.

**Comment by bsteinb** [comment]
Should be fixed by #18.
